### PR TITLE
HDDS-6768. Add Volume and Bucket ID to the key path for FSO Objects

### PR DESCRIPTION
**Pull request by nandakumar131**
## What changes were proposed in this pull request?
Adding Volume and Bucket ID to the key path for FSO Objects

## What is the link to the Apache JIRA
HDDS-6768

## How was this patch tested?
Modified/updated the existing unit/integration tests.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 1]
```diff
@@ -2298,25 +2309,27 @@ private void slimLocationVersion(OmKeyInfo... keyInfos) {
   }
 
   @Override
-  public OmKeyInfo getPendingDeletionDir() throws IOException {
-    OmKeyInfo omKeyInfo = null;
+  public Table.KeyValue<String, OmKeyInfo> getPendingDeletionDir()
+          throws IOException {
+    // TODO: Make the return type as OmDirectoryInfo after adding
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:2314]: Please mention the jira(create if not) to modify OmDirectoryInfo. Like we discussed the splitting logic is error prone and will refactor immediately after this patch.

**Review by rakeshadr — COMMENTED** [review]
Thanks @nandakumar131 for heavy lifting the changes in single patch.  Added minor comments.

Need to update the documentation section: [PrefixFSO.md](https://github.com/apache/ozone/blob/master/hadoop-hdds/docs/content/feature/PrefixFSO.md). This can be done in a separate jira task as the image has to be modified.

**Comment by rakeshadr** [comment]
@nandakumar131 Due to the recent commit changes, following test cases are failing
```
org.apache.hadoop.ozone.om.response.key.TestOMOpenKeysDeleteResponse
org.apache.hadoop.ozone.om.request.key.TestOMOpenKeysDeleteRequest
```

**Comment by nandakumar131** [comment]
Created [HDDS-6797](https://issues.apache.org/jira/browse/HDDS-6797) for [PrefixFSO.md](https://github.com/apache/ozone/blob/master/hadoop-hdds/docs/content/feature/PrefixFSO.md) update.

File: hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto [hunk 2]
```diff
@@ -183,7 +183,7 @@ message OMRequest {
 
   optional RevokeS3SecretRequest            RevokeS3SecretRequest          = 93;
 
-  optional PurgePathsRequest                purgePathsRequest              = 94;
```
Comment (prashantpogde) [inline hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:186]: Is there a reason for this rename ?

File: hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto [hunk 3]
```diff
@@ -261,7 +261,7 @@ message OMResponse {
 
   optional ListTrashResponse                  listTrashResponse            = 91;
   optional RecoverTrashResponse               RecoverTrashResponse         = 92;
-  optional PurgePathsResponse                 purgePathsResponse           = 93;
+  optional PurgeDirectoriesRequest            purgeDirectoriesResponse     = 93;
```
Comment (prashantpogde) [inline hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:264]: s/PurgeDirectoriesRequest/purgeDirectoriesResponse ??

File: hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto (hunk 2 above)
Comment (nandakumar131) [reply hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:186]: PurgePathRequest is refactored into an internal message inside PurgeDirectoryRequest to include volumeId and bucketId

File: hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto (hunk 3 above)
Comment (nandakumar131) [reply hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:264]: Thanks for catching this.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java [hunk 4]
```diff
@@ -146,7 +146,12 @@ public Collection<OzoneFileStatus> listStatusFSO(OmKeyArgs args,
 
       // fetch the db key based on parent prefix id.
       long id = getId(fileStatus, omBucketInfo);
-      dbPrefixKey = metadataManager.getOzonePathKey(id, "");
+      final long volumeId = metadataManager.getVolumeId(
```
[the remark this answers is not in the document]
Comment (nandakumar131) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java:149]: Refactored the code such that we don't need lock anymore to get _volumeId_ and _bucketId_.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java [hunk 5]
```diff
@@ -146,7 +146,12 @@ public Collection<OzoneFileStatus> listStatusFSO(OmKeyArgs args,
...
+              omBucketInfo.getVolumeName());
+      final long bucketId = metadataManager.getBucketId(
```
[the remark this answers is not in the document]
Comment (nandakumar131) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java:151]: Fixed.

**Pull request merged**